src/nrtk_explorer/app/image_ids.py

```diff
@@ -0,0 +1,15 @@
+from typing import Union
+
+DatasetId = str
+SourceImageId = str
+TransformedImageId = str
+ImageId = Union[SourceImageId, TransformedImageId]
+ResultId = str
```

Comment:
bring this comment back. I originally commented that typing primitive types seemed overkill

# response

This is a Domain Modeling trick: https://www.thoughtworks.com/en-us/insights/blog/microservices/domain-modeling-algebraic-data-types-pt2

Its useful because if you pass a DatasetId to image_id_to_result_id that would be a semanic problem. In Typescript you can enfoce this with "branded" types but I don't know how to do that with MyPy. But we don't have to do this.

I see, the usefulness of this is that our Image ids are actually different pseudo-types (by convention) we do this by prefixing the str. This changes types tries to project this using types.

+ This seems to ensure that the correct "type" is passed (our [not] prefixed str)
+ Bad thing is that we do not really enforce that the string is in a particular format corresponding to the type: imagine calling `image_id_to_dataset_id("transformed_img_2131").
+ This gives the false impression of type safety while not being the case.
+ One more lookup for checking which type really is the variable

I think that I am missing something I want to learn the benefit of this.

Reply:
I swear it is a thing 😅

https://github.com/Kitware/SMQTK-Classifier/blob/master/smqtk_classifier/interfaces/classification_element.py#L9-L14

Reply:
Thanks for the link, still I cannot see the usefulness of this. I come from the C++ realm where typing is mandatory and in generic programming strongly recommended (traits and concepts), we create types to optimize, reduce overloads generations and safer programming, the idea for safety being that we restrict functions/classes to use a particular set of types [that can have traits, such as iterable/hashable...] (as in here), creating new types that are alias to primitives is it only useful when we expect that primitive to be changed (ex: in diff platforms|archs).

Reply:
Then str it is.